src/test/python/deep_qa_test/layers/backend/batch_dot_test.py

```diff
@@ -0,0 +1,32 @@
+# pylint: disable=no-self-use,invalid-name
+import numpy
+from numpy.testing import assert_almost_equal
+import keras.backend as K
+
+from deep_qa.layers.backend.batch_dot import BatchDot
+
+
+class TestMaskedBatchDotLayer:
+    def test_compute_mask(self):
+        tensor_a = K.variable(numpy.array([[[0.3, 0.1], [0.4, 0.2], [0.8, 0.1]],
+                                           [[0.3, 0.1], [0.4, 0.2], [0.8, 0.1]]]))
+        mask_a = K.variable(numpy.array([[1, 0, 1], [1, 1, 0]]))
+        tensor_b = K.variable(numpy.array([[[0.2, 0.6], [0.4, 0.3], [0.5, 0.7], [0.1, .6]],
+                                           [[0.2, 0.6], [0.4, 0.3], [0.5, 0.7], [0.1, .6]]]))
+        mask_b = K.variable(numpy.array([[0, 1, 1, 1], [1, 0, 1, 1]]))
+        calculated_mask = K.eval(BatchDot().compute_mask([tensor_a, tensor_b],
+                                                         [mask_a, mask_b]))
+        assert_almost_equal(calculated_mask, numpy.array([[[0.0, 1.0, 1.0, 1.0],
```

Comment:
I feel like this would be easier to read and write if you did something like

```python
mask_a = numpy.ones(shape)
mask_a[0][1] = 0
mask_a[1][2] = 0
...
assert_almost_equal(calculated_mask[0][1], numpy.zeros(shape))
assert_almost_equal(calculated_mask[1][2], numpy.zeros(shape))
...
```

The reason I suggest this is because it'd be nice to have some more tests.  Switching the order of the masks above makes me nervous, because I'm not sure the mask shape is going to match the output shape anymore.  Can you add a test where you switch the order of the inputs, and make sure you get the right output?  Also, it'd be nice to have some tests with higher-order tensors, to be sure this still works.  And even though `K.batch_dot` is well tested, it'd still be nice to have one simple test with order 4 or so tensors, so that we can be sure that our layer does what we think it does, and that it breaks if keras changes underneath us.

Specifically, one test that does something like the following would probably be enough:

```python
batch_size = 3
tensor_a = numpy.random.randInt(7, shape=(batch_size, 4, 5))
tensor_b = numpy.random.randInt(7, shape=(batch_size, 6, 2, 5))
# Manually set some values to 1 here, which will be masked later (1 and not 0 so that masked values are still non-zero in the output)
input_tensor_a = Input(shape=(4, 5))
masked_tensor_a = Masking(input_tensor_a, mask_value=1)
input_tensor_b = Input(shape=(6, 2, 5))
masked_tensor_b = Masking(input_tensor_b, mask_value=1)
a_dot_b = BatchDot()([masked_tensor_a, masked_tensor_b])
a_dot_b_mask = OutputMask()(a_dot_b)
model = Model(inputs=[input_tensor_a, input_tensor_b], output=[a_dot_b, a_dot_b_mask])
a_dot_b_tensor, mask_tensor = model.predict([tensor_a, tensor_b])
# Test that the dot happened like you expected, something like this:
for i in range(batch_size):
    assert_almost_equal(a_dot_b_tensor[i], numpy.tensordot(tensor_a[i], tensor_b[i], 1))
    # Check that the values in the output mask are 0 where you set values to 1 above.
a_dot_b_tensor, mask_tensor = model.predict([tensor_b, tensor_a])
# Repeat the asserts above, but flipped.
```